src/web/mdns-publisher/mdns_publisher.cpp

```diff
@@ -28,267 +28,277 @@
+}
+
+void Publisher::HandleEntryGroupStart(AvahiEntryGroup *aGroup, AvahiEntryGroupState aState,
+                                      AVAHI_GCC_UNUSED void *aUserData)
+{
+    Publisher *publisher = static_cast<Publisher *>(aUserData);
+
+    publisher->HandleEntryGroupStart(aGroup, aState);
+}
 
 void Publisher::Free(void)
 {
-    if (mServer)
+    if (mClient != NULL)
     {
-        avahi_server_free(mServer);
+        avahi_client_free(mClient);
+        mClient = NULL;
     }
 
-    if (mSimplePoll)
+    if (mSimplePoll != NULL)
     {
         avahi_simple_poll_free(mSimplePoll);
+        mSimplePoll = NULL;
     }
 
-    avahi_free(mServiceName);
+    if (mServiceName)
+    {
+        avahi_free(mServiceName);
+        mServiceName = NULL;
+    }
+
+    if (mClientGroup != NULL)
+    {
+        avahi_entry_group_free(mClientGroup);
+    }
 }
 
-int Publisher::CreateService(AvahiServer *aServer)
+Publisher::Publisher(void) :
+    mClientGroup(NULL),
+    mSimplePoll(NULL),
+    mClient(NULL),
+    mPort(0),
+    mServiceName(NULL),
+    mNetworkNameTxt(NULL),
+    mExtPanIdTxt(NULL),
+    mType(NULL)
 {
-    int ret = kMdnsPublisher_OK;
+}
 
-    assert(aServer);
+int Publisher::CreateService(AvahiClient *aClient)
+{
+    char *serviceName;
+    int   ret = kMdnsPublisher_OK;
+
+    assert(aClient);
 
-    if (!mGroup)
+    if (!mClientGroup)
     {
-        VerifyOrExit((mGroup = avahi_s_entry_group_new(aServer, EntryGroupCallback, NULL)) != NULL,
-                     ret = kMdnsPublisher_FailedCreateGoup);
+        mClientGroup = avahi_entry_group_new(aClient, HandleEntryGroupStart, this);
+        VerifyOrExit(mClientGroup != NULL, ret = kMdnsPublisher_FailedCreateGoup);
     }
 
-    syslog(LOG_ERR, "Adding service '%s'", mServiceName);
+    if (avahi_entry_group_is_empty(mClientGroup))
+    {
+        syslog(LOG_ERR, "Adding service '%s'", mServiceName);
 
-    VerifyOrExit(avahi_server_add_service(aServer, mGroup, AVAHI_IF_UNSPEC,
-                                          AVAHI_PROTO_UNSPEC, (AvahiPublishFlags) 0,
-                                          mServiceName, mType, NULL, NULL, mPort,
-                                          mNetworkNameTxt, mExtPanIdTxt, NULL) == 0,
-                 ret = kMdnsPublisher_FailedAddSevice);
+        VerifyOrExit(avahi_entry_group_add_service(mClientGroup, AVAHI_IF_UNSPEC,
+                                                   AVAHI_PROTO_UNSPEC, (AvahiPublishFlags) 0,
+                                                   mServiceName, mType, NULL, NULL, mPort,
+                                                   mNetworkNameTxt, mExtPanIdTxt, NULL) == 0,
+                     ret = kMdnsPublisher_FailedAddSevice);
 
-    syslog(LOG_INFO, " Service Name: %s Port: %d Network Name: %s Extended Pan ID: %s ",
-           mServiceName, mPort, mNetworkNameTxt, mExtPanIdTxt);
+        syslog(LOG_INFO, " Service Name: %s  Port: %d  Network Name: %s  Extended Pan ID: %s ",
```

Comment:
unnecessary space at end of syslog format string. And is the preceding space in intended?